### PR TITLE
feat(SFINT-2829): Create a Copy to clipboard component

### PR DESCRIPTION
**Pull request by jeremierobert-coveo**
# Options
```typescript
    /**
     * The icon that the ResultAction will display.
     * If text is provided, the button will contain that text.
     * If the HTML of an SVG image is provided, that image will be displayed in the button.
     */
    icon?: string;

    /**
     * The tooltip that displays on hovering the ResultAction.
     */
    tooltip?: string;

    /**
     * Specifies the template that will be used for the copy to clipboard.
     *
     * Default value is `${title}\n${ClickUri}`.
     */
    template?: string;
```

# UA event
```
type: Click
cause: copyToClipboard
```

# Video
https://share.vidyard.com/watch/K5stZRJGtgVgewquEkRxHw?

### CODE REVIEW

**Comment by nathanlb** [comment]
Can we see a paste? :D

**Comment by jeremierobert-coveo** [comment]
Look at the video

**Comment by jeremierobert-coveo** [comment]
Oh the video dos not show my notepad =="

File: src/components/CopyPaste/CopyPaste.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        });
+    }
+
+    protected doAction() {
+        this.usageAnalytics.logClickEvent({ name: 'copyToClipboard', type: 'Click' }, {}, this.result, this.element);
+        this.copyToClipboard(StringUtils.buildStringTemplateFromResult(this.options.template, this.result));
+    }
+
+    /**
+     * Sadly that's the only way of doing in in IE11 and in lockerservice.
```
Comment (lbergeron) [inline src/components/CopyPaste/CopyPaste.ts]: This comment could be removed. The `copyToClipboardIE11` function name is clear enough.

File: src/components/CopyPaste/CopyPaste.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+     */
```
Comment (louis-bompart) [inline src/components/CopyPaste/CopyPaste.ts]: Which part is specific to LockerService?
How would it be if it wasn't for LockerService?

My point is that we should not have LockerService/Salesforce specific code in this repo.

File: src/components/CopyPaste/Strings.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import { Translation, Language } from '../../utils/translation';
+
+Translation.register(Language.English, {
+    CopyPaste_copy: 'Copy',
+    CopyPaste_copied: 'Copied !'
```
Comment (louis-bompart) [inline src/components/CopyPaste/Strings.ts]:
```suggestion
    CopyPaste_copied: 'Copied!'
```

https://www.btb.termiumplus.gc.ca/tcdnstyl-chap?lang=eng&lettr=chapsect7&info0=7

Comment (lbergeron) [inline src/components/CopyPaste/Strings.ts]: Remove whitespace before the exclamation mark.

File: src/sass/ResultActionMenu.scss [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+// Fix for quickview tooltip
+.CoveoQuickview.coveo-result-actions-menu-menu-item {
+    @extend %not-selectable;
+
+    > div {
+        height: $menu-height;
+        width: $menu-height;
+    }
+
+    // Kill poper.js
```
Comment (louis-bompart) [inline src/sass/ResultActionMenu.scss]:
```suggestion
    // Kill popper.js
```

File: src/sass/ResultActionMenu.scss [hunk 5]
```diff
@@ -0,0 +1,152 @@
+@import './Variables.scss';
```
Comment (louis-bompart) [inline src/sass/ResultActionMenu.scss]: Why is that suddenly included? It seems out of scope

File: tests/components/CopyPaste/CopyPaste.spec.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
+import { IQueryResult, $$, l } from 'coveo-search-ui';
+import { createSandbox, SinonSandbox, SinonStub } from 'sinon';
+import { Mock, Fake } from 'coveo-search-ui-tests';
+import { CopyPaste } from '../../../src/components/CopyPaste/CopyPaste';
+
+describe('ResultAction', () => {
```
Comment (louis-bompart) [inline tests/components/CopyPaste/CopyPaste.spec.ts]:
```suggestion
describe('CopyPaste ResultAction', () => {
```

File: src/components/CopyPaste/CopyPaste.ts (hunk 2 above)
Comment (jeremierobert-coveo) [reply src/components/CopyPaste/CopyPaste.ts]: There is no locker service specific code.

File: src/sass/ResultActionMenu.scss (hunk 5 above)
Comment (jeremierobert-coveo) [reply src/sass/ResultActionMenu.scss]: Because the result action css was broken.

File: src/components/ResultAction/ResultAction.ts [hunk 7]
```diff
@@ -44,6 +42,10 @@ export abstract class ResultAction extends Component {
         tooltip: ComponentOptions.buildStringOption()
     };
 
+    public tooltip: HTMLElement;
```
Comment (lbergeron) [inline src/components/ResultAction/ResultAction.ts]: We shouldn't expose the tooltip element. It is breaking component encapsulation.

File: src/sass/ResultActionMenu.scss (hunk 5 above)
Comment (louis-bompart) [reply src/sass/ResultActionMenu.scss]: That's seems like another issue then, please split :D

File: src/components/CopyPaste/CopyPaste.ts (hunk 2 above)
Comment (louis-bompart) [reply src/components/CopyPaste/CopyPaste.ts]: Then the comment needs to be updated

File: src/sass/ResultActionMenu.scss (hunk 5 above)
Comment (jeremierobert-coveo) [reply src/sass/ResultActionMenu.scss]: No that css is require for Copy Paste.

File: src/sass/ResultActionMenu.scss [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+// Fix for quickview tooltip
+.CoveoQuickview.coveo-result-actions-menu-menu-item {
+    @extend %not-selectable;
+
+    > div {
+        height: $menu-height;
+        width: $menu-height;
+    }
+
+    // Kill poper.js
```
Comment (lbergeron) [inline src/sass/ResultActionMenu.scss]: Can you use `// Override poper.js` instead? It would sound less radical ;)

File: src/sass/ResultActionMenu.scss (hunk 5 above)
Comment (louis-bompart) [reply src/sass/ResultActionMenu.scss]: Explain please, I don't understand why a new ResultAction requires changes in the menu, changes in the QuickviewAction etc.
I'm still not convinced that this is in the scope.

File: src/components/CopyPaste/CopyPaste.ts (hunk 2 above)
Comment (jeremierobert-coveo) [reply src/components/CopyPaste/CopyPaste.ts]: Why ?

File: src/sass/ResultActionMenu.scss (hunk 5 above)
Comment (jeremierobert-coveo) [reply src/sass/ResultActionMenu.scss]: Because of the existing css result actions look out of place when compared with quickview result actions.

File: tests/components/CopyPaste/CopyPaste.spec.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        element = document.createElement('div');
+        document.body.append(element);
+        testOptions = new Mock.AdvancedComponentSetupOptions(element);
+
+        result.title = 'some title';
+        result.clickUri = 'some uri';
+
+        componentSetup = Mock.advancedResultComponentSetup(CopyPaste, result, testOptions);
+        testComponent = componentSetup.cmp;
+        writeTextStub = sandbox.stub(navigator.clipboard, 'writeText').resolves();
```
Comment (lbergeron) [inline tests/components/CopyPaste/CopyPaste.spec.ts]: This does not test the IE11 code path. It would be nice to have a test case that covers IE11.

File: src/components/CopyPaste/CopyPaste.ts (hunk 2 above)
Comment (louis-bompart) [reply src/components/CopyPaste/CopyPaste.ts]: It's literally written **lockerservice** in the comment, if there's no part that is specific in lockerservice, then it's only a IE11 issue and thus the `and in lockerservice` is not needed

File: src/sass/ResultActionMenu.scss (hunk 5 above)
Comment (louis-bompart) [reply src/sass/ResultActionMenu.scss]: And that's another issue :)

File: src/components/CopyPaste/CopyPaste.ts (hunk 2 above)
Comment (jeremierobert-coveo) [reply src/components/CopyPaste/CopyPaste.ts]: It's also a lockerservice issue because the Clipboard AP is not available in locker service as in IE11 end of story and locker service needs to be mention because if we remove support for IE11 we still need this code for locker service.

File: src/sass/ResultActionMenu.scss (hunk 5 above)
Comment (jeremierobert-coveo) [reply src/sass/ResultActionMenu.scss]: So you want me to ship an half feature instead of a complete one ?

File: src/components/CopyPaste/CopyPaste.ts (hunk 2 above)
Comment (mikegron) [reply src/components/CopyPaste/CopyPaste.ts]: cause it mentions lockerservice? 📦

File: tests/components/CopyPaste/CopyPaste.spec.ts [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+    describe('on click', () => {
+        it('should paste the result in the clipboard', async () => {
+            testComponent.element.click();
+            expect(writeTextStub.called).toBeTrue();
+        });
+    });
+
+    describe('on mouseleave', () => {
+        it('should change the tooltip text for "Copy"', () => {
+            testComponent.tooltip.innerText = 'patate';
```
Comment (lbergeron) [inline tests/components/CopyPaste/CopyPaste.spec.ts]: I think you souldn't expose `.tooltip`. I guess you should validate this behavior using a functional test instead. If you prefer to use a unit test, maybe we could expose an event instead.

File: src/components/CopyPaste/CopyPaste.ts (hunk 2 above)
Comment (jeremierobert-coveo) [reply src/components/CopyPaste/CopyPaste.ts]: @mikegron you respond at which question ?

File: src/components/CopyPaste/CopyPaste.ts [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+
+    /**
+     * Specifies the template that will be used for the copy to clipboard.
+     *
+     * Default value is `${title}\n${ClickUri}`.
+     */
+    template?: string;
+}
+
+export class CopyPaste extends ResultAction {
```
[the remark this answers is not in the document]
Comment (jeremierobert-coveo) [reply src/components/CopyPaste/CopyPaste.ts]: ClickUri is just an option you can change this component to copy the excert to clipboard so it's not bound to a field

File: src/components/CopyPaste/CopyPaste.ts (hunk 2 above)
Comment (louis-bompart) [reply src/components/CopyPaste/CopyPaste.ts]: So, the clipboard API does not work in LockerService even in Chrome ?

File: tests/components/CopyPaste/CopyPaste.spec.ts (hunk 9 above)
Comment (jeremierobert-coveo) [reply tests/components/CopyPaste/CopyPaste.spec.ts]: Sadly you can't. (Trust me I try)

File: tests/components/CopyPaste/CopyPaste.spec.ts (hunk 10 above)
Comment (jeremierobert-coveo) [reply tests/components/CopyPaste/CopyPaste.spec.ts]: You can't use Functionnal tests because of some issues in cypress regarding clipboard and hovering

File: src/components/CopyPaste/CopyPaste.ts (hunk 2 above)
Comment (jeremierobert-coveo) [reply src/components/CopyPaste/CopyPaste.ts]: Exacly

Comment (jeremierobert-coveo) [reply src/components/CopyPaste/CopyPaste.ts]: image

File: src/sass/ResultActionMenu.scss (hunk 5 above)
Comment (louis-bompart) [reply src/sass/ResultActionMenu.scss]: I want to ship the strict minimum for the feature, this doesn't seems like it.
Can you indicate in this file which CSS rule concerns directly and exclusively the CopyAndPaste component?

File: src/components/CopyPaste/CopyPaste.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (lbergeron) [reply src/components/CopyPaste/CopyPaste.ts]: I see your point. I think `CopyResult` would be a better name name. In my opinion, `Paste` is irrelevant as the component only copies data. I pastes nothing.

File: src/components/CopyPaste/CopyPaste.ts [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+        public element: HTMLElement,
+        public options: ICopyPasteOptions,
+        public bindings?: IResultsComponentBindings,
+        public result?: IQueryResult
+    ) {
+        super(element, ComponentOptions.initComponentOptions(element, CopyPaste, options), bindings, result);
+
+        super.init();
+
+        this.bind.on(this.element, 'mouseleave', (event: MouseEvent) => {
```
Comment (mikegron) [inline src/components/CopyPaste/CopyPaste.ts]: ont peux pas set le innerText au debut ? sa me fait bizzare de setté ca dans un event mais bon

File: src/components/CopyPaste/CopyPaste.ts (hunk 2 above)
Comment (lbergeron) [reply src/components/CopyPaste/CopyPaste.ts]: Why not name the method `copyToClipboardFallback` instead? It would include IE11 and LockerService without naming them explicitely.

File: src/components/CopyPaste/CopyPaste.ts [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    /**
+     * Sadly that's the only way of doing in in IE11 and in lockerservice.
+     */
+    private copyToClipboard(text: string) {
+        if (navigator && navigator.clipboard) {
+            navigator.clipboard
+                .writeText(text)
+                .then(() => (this.tooltip.innerText = l('CopyPaste_copied')), err => this.logger.error('Copy to clipboard failed.', text, err));
```
Comment (mikegron) [inline src/components/CopyPaste/CopyPaste.ts]: l'error devrais pas être dans un `.catch()` pour simplicité ?

File: src/components/CopyPaste/CopyPaste.ts [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+        } else {
+            this.copyToClipboardIE11(text);
+        }
+    }
+
+    private copyToClipboardIE11(text: string) {
```
Comment (mikegron) [inline src/components/CopyPaste/CopyPaste.ts]: mes sympathies 🥀

File: src/sass/ResultActionMenu.scss (hunk 5 above)
Comment (mikegron) [reply src/sass/ResultActionMenu.scss]: I agree with jeremie

File: src/sass/ResultActionMenu.scss [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+}
+
+%select-all {
+    -webkit-user-select: all;
+    -moz-user-select: all;
+    -ms-user-select: all;
+    user-select: all;
+}
+
+.CoveoResultActionsMenu {
```
Comment (mikegron) [inline src/sass/ResultActionMenu.scss]: Si je comprend bien c'est un copy-paste du scss de search ui avec des petites modifications ? C'est pas possible d'avoir juste les modifs pour pas avoir ce gros fichier?

File: tests/components/CopyPaste/CopyPaste.spec.ts (hunk 9 above)
Comment (lbergeron) [reply tests/components/CopyPaste/CopyPaste.spec.ts]: Did you try spying on `document.execCommand`? Or we could add an event like `onCopy` that would contain the copied information. This way we would be able to easily validate that the right data is sent without having to deal directly with the browser.

File: src/sass/ResultActionMenu.scss (hunk 15 above)
Comment (jeremierobert-coveo) [reply src/sass/ResultActionMenu.scss]: Pas dutout, c'est pas du jsui c'est de SFINT. Une belle mise a jour

File: src/components/CopyPaste/CopyPaste.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (jeremierobert-coveo) [reply src/components/CopyPaste/CopyPaste.ts]: What about `CopyToClipboard` ?

File: src/components/CopyPaste/CopyPaste.ts (hunk 1 above)
Comment (jeremierobert-coveo) [reply src/components/CopyPaste/CopyPaste.ts]: will change the function name for `copyToClipboardFallback` so I'll keep this comment.

File: src/components/CopyPaste/CopyPaste.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (lbergeron) [reply src/components/CopyPaste/CopyPaste.ts]: Yes I'm fine with that!

File: src/components/CopyPaste/CopyPaste.ts (hunk 12 above)
Comment (jeremierobert-coveo) [reply src/components/CopyPaste/CopyPaste.ts]: Nope tu changer le texte de copy a copied!

File: src/components/CopyToClipboard/CopyToClipboard.ts [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+
+export class CopyToClipboard extends ResultAction {
+    static ID = 'CopyToClipboard';
+
+    /**
+     * The possible options for _ResultAction_.
+     * @componentOptions
+     */
+    static options: ICopyToClipboardOptions = {
+        icon: ComponentOptions.buildStringOption({ defaultValue: copyToClipboard }),
```
Comment (louis-bompart) [inline src/components/CopyToClipboard/CopyToClipboard.ts]: Too much breaklines imho here. Not blocking for review tho.

File: src/components/CopyToClipboard/CopyToClipboard.ts [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+    protected doAction() {
+        this.usageAnalytics.logClickEvent({ name: 'copyToClipboard', type: 'Click' }, {}, this.result, this.element);
+        this.copyToClipboard(StringUtils.buildStringTemplateFromResult(this.options.template, this.result));
+    }
+
+    private copyToClipboard(text: string) {
+        if (navigator && navigator.clipboard) {
+            navigator.clipboard
+                .writeText(text)
+                .then(() => (this.element.querySelector<HTMLElement>('.coveo-caption-for-icon').innerText = l('CopyToClipboard_copied')))
```
Comment (louis-bompart) [inline src/components/CopyToClipboard/CopyToClipboard.ts]: async/await instead?

File: src/components/ResultAction/ResultAction.ts [hunk 18]
```diff
@@ -44,6 +42,8 @@ export abstract class ResultAction extends Component {
         tooltip: ComponentOptions.buildStringOption()
     };
 
+    private isInitialized = false;
```
Comment (louis-bompart) [inline src/components/ResultAction/ResultAction.ts]: Mhh, revert the change on this file I'd say?

**Comment by jeremierobert-coveo** [comment]
All comments resolve and change implemented

File: tests/components/UserActions/QueryList.spec.ts [hunk 19]
```diff
@@ -269,30 +269,6 @@ describe('QueryList', () => {
             });
         });
 
-        it('should log a search event with "userActionsSubmit" as search cause', () => {
```
Comment (lbergeron) [inline tests/components/UserActions/QueryList.spec.ts]: Why are you removing this test?

Comment (jeremierobert-coveo) [reply tests/components/UserActions/QueryList.spec.ts]: usageAnalytics is no more mockable by updating the jsui

Comment (lbergeron) [reply tests/components/UserActions/QueryList.spec.ts]: Upgrading from `2.0.342` to `2.8521.7` is a pretty big leap forward indeed. Ok then.
